### PR TITLE
Update review policy

### PR DESCRIPTION
**Pull request by michaelwoerister**
Discussing this was proposed in: https://github.com/rust-lang/compiler-team/issues/444. The actual discussion took place [on zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bsteering.20meeting.5D.202021-07-16.20compiler-team.23444).

[Rendered](https://github.com/michaelwoerister/rust-forge/blob/review-policy/src/compiler/reviews.md)

@rust-lang/compiler: Please leave comments about anything that you'd still like to have discussed.

### CODE REVIEW

**Comment by michaelwoerister** [comment]
> I think it'd be good to add some section on "not assigned reviews" and best practices there -- both in terms of how to indicate you aren't the final reviewer to the author, but also whether e.g. it's acceptable to r- someone's PR asking them to cleanup commit history / fix the description / etc.

@Mark-Simulacrum, where would you like this to be added? Do you want to draft the concrete text?

**Comment by michaelwoerister** [comment]
> I voted :writing: on this because I feel like if you're not feeling comfotable with reviewing a PR, it is unlikely you'll be able to accurately gauge the scope of the PR and/or its contentiousness.

@nagisa, do you want to elaborate on that?

File: src/compiler/reviews.md
Comment (michaelwoerister) [inline src/compiler/reviews.md:146]: @Mark-Simulacrum, I think you wanted to discuss the wording here. Do you have specific comments?

File: src/compiler/reviews.md
Comment (michaelwoerister) [inline src/compiler/reviews.md:105]: @pnkfelix & @wesleywiser,  you both signaled that you wanted to discuss the wording of this paragraph.

File: src/compiler/reviews.md
Comment (michaelwoerister) [inline src/compiler/reviews.md:96]: @pnkfelix, you signaled that you wanted to discuss the wording of this paragraph.

File: src/compiler/reviews.md
Comment (oli-obk) [inline src/compiler/reviews.md:18]: one additional point that I have had trouble with (both as a reviewer and as someone looking for a reviewer) is that all of the points below were satisfied, but the reviewer was already swamped with reviews. What I'm trying to say is essentially "what do we do with PRs where no one can review it?".

File: src/compiler/reviews.md
Comment (jackh726) [inline src/compiler/reviews.md:89]: I think this will probably come up with the discussion on getting more people in the review queue, but I think we should also emphasize that even if you can't do the *final* review for whatever reason, it's always okay and *helpful* to take a first glance and give comments. It's a win-win all around: contributors might get at least some feedback sooner; the initial reviewer can at least gets eyes on code that they might be new to; the final reviewer might have less things that need to be said/addressed when they get to their review.

**Review by jackh726 — APPROVED** [review]

File: src/compiler/reviews.md
Comment (tmandry) [reply src/compiler/reviews.md:89]: +1!

Comment (tmandry) [reply src/compiler/reviews.md:89]: Also, can we dedicate a Zulip stream for this? That seems more approachable than asking specific compiler team members (which member to ask?)

File: src/compiler/reviews.md
Comment (oli-obk) [reply src/compiler/reviews.md:18]: That indeed doesn't belong to this document.

I was rather thinking along the lines of "can we reject a PR due to insufficient reviewer capacity instead of letting it die slowly by not reviewing it"

File: src/compiler/reviews.md
Comment (michaelwoerister) [reply src/compiler/reviews.md:35]: Yes, we'll have to iterate on that.

File: src/compiler/reviews.md
Comment (michaelwoerister) [reply src/compiler/reviews.md:89]: A Zulip stream sounds like a good idea. Any suggestions for the name of such a stream?

File: src/compiler/reviews.md
Comment (michaelwoerister) [reply src/compiler/reviews.md:18]: That's a good question -- and I have no good answer for it `:)` But it sounds like it would be a good bullet point for the guidance section.

Comment (michaelwoerister) [reply src/compiler/reviews.md:18]: I've opened https://github.com/rust-lang/compiler-team/issues/448 for keeping track of this question.

File: src/compiler/reviews.md
Comment (rylev) [inline src/compiler/reviews.md:22]: Nit: do we want to rephrase this in terms of what the PR author must provide? The other points are about obligations of the reviewer, but this point is about what a pull request must provide. Obviously, pull requests don't have the ability to provide anything (they're just UI on GitHub) - it is the responsibility of the author to provide these things *in* a pull request description.

File: src/compiler/reviews.md
Comment (rylev) [inline src/compiler/reviews.md:23]: Besides contention, it would also be good for the author to point the reviewer to anything the author was unsure about (e.g., solutions that didn't seem ideal, compromises that needed to be made, etc.). In other words, what are the "known unknowns" of this PR from the perspective of the author? Perhaps this can be added to the checklist?

File: src/compiler/reviews.md
Comment (rylev) [inline src/compiler/reviews.md:30]: nit: not only timely, but also sufficiently thorough: a quick but not sufficiently thorough review is not what we want either.

File: src/compiler/reviews.md
Comment (wesleywiser) [reply src/compiler/reviews.md:89]: `#t-compiler/reviews` perhaps?

File: src/compiler/reviews.md
Comment (michaelwoerister) [reply src/compiler/reviews.md:108]: I think the heading of the paragraph should be a description of the specific problem one encounters. At least that's what I've been going for for the other sections. Does that make sense? I'm happy to rephrase the heading if that makes it clearer though.

File: src/compiler/reviews.md
Comment (michaelwoerister) [reply src/compiler/reviews.md]: Done

File: src/compiler/reviews.md
Comment (michaelwoerister) [reply src/compiler/reviews.md]: Done

File: src/compiler/reviews.md
Comment (michaelwoerister) [reply src/compiler/reviews.md]: Done.

File: src/compiler/reviews.md
Comment (michaelwoerister) [reply src/compiler/reviews.md]: I switched this to the Zulip stream version.

File: src/compiler/reviews.md
Comment (Mark-Simulacrum) [reply src/compiler/reviews.md:149]: I wanted to suggest a few changes:

* Replace the asking on the Zulip stream with some kind of more transactional process (e.g., handover to Felix/Wesley for re-assignment -- something that more fully moves things out of the current reviewer's hands). Maybe indicating the PR is S-waiting-on-team. Questions on Zulip can easily get lost I think, and bouncing around seems ungreat. Maybe we can I-nominate for T-compiler meetings with the explicit goal of assigning a reviewer, but those meetings don't usually have full attendance so it may not be the best venue either.

**Comment by LeSeulArtichaut** [comment]
(rust-lang/highfive#350 has been merged)

**Comment by davidtwco** [comment]
Little late to the discussion here, updated review policy looks good to me, but wanted to suggest another edge case that we should document in our review policy: a pull request which makes changes to support the use of rustc internals for some external project - these aren't common but I've seen it happen a few times - do we accept these changes if they're small and otherwise improvements, do we reject these unless it's a mature and released project, etc?

**Comment by oli-obk** [comment]
I think we should allow changes making things public, cleaning up things or making them more general, as long as the owners of the compiler region agree (so just assign to them). As a concrete example: if someone is using the mir interpreter, and they want to make something public, it is likely not a problem, but there are some functions that are module- or crate-private on purpose, as they uphold invariants within the mir interpreter. So basically, just assign such PRs to the relevant people (usually they get pinged anyway due to having told highfive that they want to get pinged on changes to these parts)

**Comment by ehuss** [comment]
@michaelwoerister I know it's been a while, are you still interested in working on this? I'm not clear if this was waiting for more updates or reviews, or if it is ready to go as-is.

**Comment by michaelwoerister** [comment]
Oh my, that was never merged, was it? Sorry for not following up!

I'll look into getting this over the finish line.

File: src/compiler/reviews.md
Comment (michaelwoerister) [reply src/compiler/reviews.md:109]: @wesleywiser, do you still think we should change the title to  "Transparent discussion and rationale".

File: src/compiler/reviews.md
Comment (michaelwoerister) [reply src/compiler/reviews.md:149]: Yes, that sounds like a good idea. Maybe:

> In such a case it is a good idea for reviewers to I-nominate the PR (if they intend to stay the main reviewer) or assign a compiler team lead to the issue and add the S-waiting-on-team label. In both cases, the PR will be brought in the weekly triage meeting.

@apiraino, what do you think of that?

File: src/compiler/reviews.md
Comment (michaelwoerister) [inline src/compiler/reviews.md:90]: This mentions a `#t-compiler/reviews` Zulip stream, which does not exist. Any objections to changing that to `#t-compiler`?

Comment (wesleywiser) [reply src/compiler/reviews.md:90]: No objection 🙂

File: src/compiler/reviews.md
Comment (michaelwoerister) [reply src/compiler/reviews.md:109]: I've changed the title "Discussion or rationale is too intransparent", should be clearer.

File: src/compiler/reviews.md
Comment (michaelwoerister) [reply src/compiler/reviews.md:151]: I've gone ahead and updated the text with the above version.

**Comment by michaelwoerister** [comment]
@wesleywiser & @davidtwco, I think the outstanding questions have been addressed (see the final two commits). I don't think there are any major changes compared to when the document was discussed originally, so I think we could merge the PR in the current state.

I did not add the point @davidtwco made in https://github.com/rust-lang/rust-forge/pull/553#issuecomment-917938159. I'd be happy to if I can get some help with what the actual text should look like.

**Comment by jieyouxu** [comment]
I tried to address @davidtwco's [concern](https://github.com/rust-lang/rust-forge/pull/553#issuecomment-917938159) about changing compiler-internal API visibility for external consumers.

I also fleshed out the review policy with some more specifics, so this will need a re-review.

cc @rust-lang/compiler because this has stalled since last year, and now has significant changes. cc compiler leads @wesleywiser and @davidtwco for approving reviews specifically.

File: src/compiler/reviews.md
Comment (SparrowLii) [inline src/compiler/reviews.md:84]: Is something missed here?

Comment (jieyouxu) [reply src/compiler/reviews.md:84]: Yes, good catch, this was supposed to be 'perf run'

Comment (jieyouxu) [reply src/compiler/reviews.md]: Fixed

File: src/compiler/reviews.md
Comment (jieyouxu) [reply src/compiler/reviews.md]: Not sure if this is still a concern.

File: src/compiler/reviews.md
Comment (jieyouxu) [reply src/compiler/reviews.md:577]: I'm inclined to add a backlink to bors docs instead

File: src/compiler/reviews.md
Comment (jieyouxu) [reply src/compiler/reviews.md:105]: Thanks, I think I wanted to remove this line but was left stray instead.

File: src/compiler/reviews.md
Comment (SparrowLii) [reply src/compiler/reviews.md:577]: That's ok

**Review by davidtwco — APPROVED** [review]
LGTM after @SparrowLii's comments addressed

File: src/compiler/reviews.md
Comment (jieyouxu) [reply src/compiler/reviews.md:577]: Linked to bors

File: src/compiler/reviews.md
Comment (jieyouxu) [reply src/compiler/reviews.md]: I'm going to mark this as resolved because it has changed a bit.

**Comment by jieyouxu** [comment]
Going to wait a bit in case there are anymore immediate feedback, otherwise merge at your discretion.

File: src/compiler/reviews.md
Comment (oli-obk) [inline src/compiler/reviews.md:67]: I'm not sure how useful this list is when it solely exists in the forge.

I think at minimum we should change the PR template to include a link to it.

In my personal ideal world we'd have this list as a list of checkboxes (reformulated so it doesn't say "Does the PR need X", but "Checked whether the PR needs X and handled it if necessary"

Comment (jieyouxu) [reply src/compiler/reviews.md:67]: This was originally phrases to be included as a PR template yeah.

**Comment by oli-obk** [comment]
I'd say we land it and edit it in place in follow-up PRs if more people comment

**Pull request merged**
